pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -430,6 +447,99 @@ public void sendAsync(Message<T> message, SendCallback callback) {
         }
     }
 
+    private boolean fillMessageSchema(MessageMetadata.Builder msgMetadataBuilder,
+                                      Schema msgSchema,
+                                      SendCallback callback) {
+        if (msgSchema == schema) {
+            schemaVersion.ifPresent(v -> msgMetadataBuilder.setSchemaVersion(ByteString.copyFrom(v)));
+            return true;
+        }
+        if (!isMultiSchemaEnabled(true)) {
+            callback.sendComplete(new PulsarClientException.InvalidMessageException(
+                    "Multiple schema disabled"));
+            return false;
+        }
+        byte[] schemaVersion;
+        try {
+            schemaVersion = schemaCache.computeIfAbsent(
+                    SchemaHash.of(msgSchema), (hash) -> {
+                        SchemaInfo schemaInfo = Optional.ofNullable(msgSchema)
+                                                        .map(Schema::getSchemaInfo)
+                                                        .filter(si -> si.getType().getValue() > 0)
+                                                        .orElse(Schema.BYTES.getSchemaInfo());
+                        try {
+                            return getOrCreateSchemaAsync(schemaInfo).get();
```

Comment:
@yittg I think this should be improved. We should avoid calling *sync* methods. 

A better implementation here is:

- if a new schema is needed, flush out all the existing pending requests to the broker. because all the existing pending requests use old schemas that are already registered in broker.
- introduce a state for the client (e.g. REGISTERING_SCHEMA). Turn the client into REGISTERING_SCHEMA state. All the requests should be added in the pending queue and wait until the schema is registered in the broker.
- after a schema is successfully registered in the broker, flush all the pending requests to the broker.

Reply:
Yeah, agree with you on executing command asynchronously. It's a better implementation, I will get it out.

Reply:
First try pushed, PTAL.